lib/sender.js

```diff
@@ -0,0 +1,119 @@
+'use strict';
+
+const net = require('net')
+const tls = require('tls')
+const { Writable } = require('stream')
+
+
+/**
+ * Event sender.
+ * Can be used as a stream, with the usual functions:
+ * sender.write(), sender.end(), etc.
+ * Implements backpressure.
+ */
+class Sender extends Writable {
+
+  /**
+   * Create the client.
+   * @param {Object} options for sender, including:
+   *  - host: the host to send events, default localhost.
+   *  - port: the destination port (mandatory).
+   *  - objectMode: whether to allow sending objects.
+   * @param {function(*)} callback (Optional) Function to invoke when the
+   * connection has been established.
+   */
+  constructor(options, callback) {
+    super(options)
+    this._options = options
+    const libnet = options.cert ? tls : net
+    this._socket = libnet.connect(options, callback)
+    this._socket.on('error', error => this.emit('error', error))
+    //this._socket.unref()
+  }
+
+  /**
+   * Send an event to the ingestion endpoint.
+   *
+   * @param {string|Buffer|Object} event message to send.
+   * If an object is passed it will be converted to JSON before sending.
+   * @param {function(*)} callback Optional function to invoke with (error).
+   * @return true if the stream accepts more events.
+   */
+  send(event, callback) {
+    const converted = this._convert(event)
+    return this._socket.write(converted, null, error => {
+      if (callback) return callback(error)
+      if (error) this.emit(error)
+    })
+  }
+
+  _convert(event) {
+    const transformed = this._transform(event)
+    //console.log('sending %s', transformed)
+    if (transformed.endsWith('\n')) return transformed
+    return transformed + '\n'
+  }
+
+  /**
+   * Generates an event that conforms to RFC 5424:
+   * https://tools.ietf.org/html/rfc5424
+   */
+  _transform(event) {
+    //console.log('Received event %j', event)
+    const bits = [
+      '<' + (this._options.priority || 13) + '>' + new Date().toISOString(),
+      this._options.localhost || 'localhost.localdomain',
+      (this._options.tag ? this._options.tag : 'my.app') + ':',
+    ]
+    if (typeof event == 'string') {
+      bits.push(event)
+    } else if (Buffer.isBuffer(event)) {
+      bits.push(String(event))
+    } else {
+      bits.push(JSON.stringify(event))
+    }
+    return bits.join(' ')
```

Comment:
since `_transform` is returning a string, it is more efficient to work with strings rather than creating an array and then joining it to a string, something like:
```
_transform(content) {
  const event = `<${(this._options.priority || 13)}>${new Date().toISOString()} ` +
    `${this._options.localhost || 'localhost.localdomain'} ` +
    `${(this._options.tag ? this._options.tag : 'my.app')}: `;

  if (typeof content == 'string') {
￼    event += content;
￼  } else if (Buffer.isBuffer(content)) {
￼    event += String(content);
￼  } else {
￼    event += JSON.stringify(content);
￼  }
  return event;
}
```

Reply:
This looks messier to me. Also, I am not sure it is more efficient as there are several intermediate concatenations in your code, something I was trying to avoid.

Reply:
what concatenations? the ones in the `if else` statements? since the method receives something that will be transformed into a string and returns a string, i guess it'll be more efficient to limit the code to the realm of strings. I'll write a benchmark to find out.

Reply:
```
string concat x 838,754 ops/sec ±0.99% (86 runs sampled)
array join x 702,763 ops/sec ±1.15% (84 runs sampled)
```
very similar. Do as you please.